mewarpx/mewarpx/sim_control.py

```diff
@@ -0,0 +1,71 @@
+""" Control code used to terminate simulation based on 
+a set of user defined criteria
+"""
+
+import logging
+import time
+
+import numpy as np
+import psutil
+
+from pywarpx import _libwarpx
+from mewarpx.mwxrun import mwxrun
+
+# Get module-level logger
+logger = logging.getLogger(__name__)
+
+class SimControl:
+    """
+    Evaluate whether to continue or terminate the simulation based on a
+    set of user defined functions and criteria
+
+    """
+    def __init__(self, criteria, criteria_args, max_loops=None, **kwargs):
+        """
+        Generate and install functions to perform after a step #.
+        Arguments:
+        
+        criteria: list of user defined functions that return True or False value
+
+        criteria_args: list of kwargs for each criteria
+        """
+
+        self.crit_list = criteria
+        self.crit_args_list = criteria_args
+        self.max_loops = max_loops
+        self.check_increment = 0
+
+    def add_checker(self, criteria=None, crit_args=None):
+        self.crit_list.append(criteria)
+        if crit_args:
+            self.crit_args_list.append(crit_args)
+        else:
+            self.crit_args_list.append({})
```

Comment:
Here you don't need the if else statement/branching logic if you set the default value of `crit_args` to `crit_args={}`